tests/pkcli_data/arghparsing/pkcli/__init__.py

```diff
@@ -0,0 +1 @@
+#
```

Comment:
These are unnecessary. You can just use `p1`. This simplifies the test data. Sorry to be pedantic, but I think for future test-learners, this is important. The argh_test.py is clear enough.

Reply:
These are good points. I only have opinions that do not mean much, but I offer them anyway. :)

`p1` could be renamed `package1` if that's clearer to you.

One question you raise is whether each case should be in its own package. That I think is perhaps valid or overkill. I don't find it useful to have empty files, e.g. argparsing/__init__.py`. This implies that there's a module named argparsing in use somewhere. That was my confusion when I first read it, and I though you were mocking it in the test.

The case name itself `test_argh_argument_parsing` is what's relevant to me. It's the first place anybody should look to see if there's a test for this case, e.g. search on `argh` in the test files for pkcli. Alternatively, it might be in a string for pkeq or something. We don't usually do that, because the test case names (to me at least) should be clear as to what they are testing. In any case, it will be in the test file. 

The test data itself is very tricky to understand without understanding the test code and the code that is being tested. It wasn't clear to me why, for example, you had a `pkeq(str(PKDict...)` until I saw the result. One could argue here (and I didn't at first :) that the use of `str` is implicitly coupled with the way pkcli main behaves, which is actually not the purpose of the test. If, for example, you simply stringified the result instead of returning a dict, then it would be clearer (at least to me) that the two values are the same. There's an awful lot to unpack here, but it's something that's subtle about testing: be careful about including too much in the test.

> I end up at functions like pkcli_test._dev. Another short name that tells me nothing. 

Unfortunately, that's just it. I wouldn't have known a good name for that function. A longer name might be fine, but there's a lot of context in these functions. The word `dev` is overloaded, I agree. `_deviance` would be clearer, but I don't know that it would explain things, but then this code would be harder to read:

```py
    _dev(rp, [], None, all_modules, capsys)
    _dev(rp, ["--help"], None, all_modules, capsys)
    _dev(rp, ["conf1"], SystemExit, r"cmd1,cmd2.*too few", capsys)
    _dev(rp, ["conf1", "-h"], SystemExit, r"\{cmd1,cmd2\}.*commands", capsys)
    if six.PY2:
        _dev(rp, ["not_found"], None, r"no module", capsys)
    else:
        _dev(rp, ["not_found"], ModuleNotFoundError, None, capsys)
    _dev(rp, ["conf2", "not-cmd1"], SystemExit, r"\{cmd1\}", capsys)
```

Perhaps a for loop would help, or not, because cases failures would be harder to track.

I'm not dying on any hills here. Just saying my opinions. The fundamental one is that testing is very hard to get right.

Reply:
> p1 could be renamed package1

1000x yes. The name package1 gets me a lot closer to realizing that the name is meaningless. I wish I knew why but with the name p1 I really don't know what to think. When names are hard to understand (p1) my first thought is that I don't understand the code. Not that the name itself is meaningless.

After learning that p1 means package1 I'm less inclined to have a separate package for the argh testing. I still like that it is separated out in the negative case. Also, the arg case feels a bit unique to me since it is a test for dealing with the peculiarities of a release of a dependency and I can imagine a future where we get rid of the test. I think separating it out makes it easier to get rid of.  But, I'm not too sold on any of those ideas.

>  The word dev is overloaded, I agree. _deviance would be clearer

Again, an emphatic yes. As you say, dev is overloaded. Deviance is a long way down the list of names I might guess. And I'd rather not guess at all. deviance actually tells me quite a bit about what is going on. It doesn't paint the whole picture but it gets me started. A relevant example: initially when I wrote the test in here I tried to use _dev because I saw other commands using it that looked like they were doing what I wanted to do. It wasn't until I ran into the assert for a return value of 1 that I realized what was going on. I think (hope) that "_deviance" would have clued me in sooner. 

> One could argue here (and I didn't at first :) that the use of str is implicitly coupled with the way pkcli main behaves, which is actually not the purpose of the test

I think switching to a string is a good idea.

Thanks again for the details.

6b14220